File: PyProject/leetcode/history/n-ary-tree-preorder-traversal.py
from typing import List


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class NAryTreePreorderTraversal:
    def preorder(self, root: 'Node') -> List[int]:
        res = []
        if root:
            res.append(root.val)
            for i in root.children:
                res.extend(self.preorder(i))
        return res

File: GoProject/leetcode/tree/n-ary-tree-preorder-traversal.go
package tree

//Definition for a Node.
type Node struct {
	Val      int
	Children []*Node
}

func helper(root *Node, res []int) []int {
	if root == nil {
		return res
	}
	res = append(res, root.Val)
	if len(root.Children) == 0 {
		return res
	}
	for _, node := range root.Children {
		res = helper(node, res)
	}
	return res
}

//func preorder(root *Node) []int {
//	return helper(root, []int{})
//}

func preorder(root *Node) (res []int) {
	var npreorder func(root *Node)
	npreorder = func(root *Node) {
		if root == nil {
			return
		}
		res = append(res, root.Val)
		for _, node := range root.Children {
			npreorder(node)
		}
	}
	npreorder(root)
	return

}
